src/main/scala/viper/gobra/translator/encodings/closures/ClosureEncoding.scala

```diff
@@ -0,0 +1,170 @@
+      val (pos, info, errT) = c.vprMeta
+      cl.unit(vpr.LocalVar(Names.closureArg, domain.vprType)(pos, info, errT))
+
+    case c@in.PureFunctionCall(Left(_), _, Some(_), _) =>
+      specs.pureClosureCall(c)(ctx)
+
+    case c@in.PureFunctionCall(Right(proxy), args, Some(spec), typ) =>
+      val member = ctx.lookup(proxy)
+      val fType = in.FunctionT(member.args.map(_.typ), member.results.map(_.typ), Addressability.rValue)
+      val fObject = in.FunctionObject(proxy, fType)(c.info)
+      specs.pureClosureCall(in.PureFunctionCall(Left(fObject), args, Some(spec), typ)(c.info))(ctx)
+
+    case c@in.PureMethodCall(recv, meth, args, Some(spec), typ) =>
+      val member = ctx.lookup(meth)
+      val mType = in.FunctionT(member.args.map(_.typ), member.results.map(_.typ), Addressability.rValue)
+      val mObject = in.MethodObject(recv, meth, mType)(c.info)
+      specs.pureClosureCall(in.PureFunctionCall(Left(mObject), args, Some(spec), typ)(c.info))(ctx)
+
+    case a: in.ClosureImplements =>
+      specs.closureImplementsExpression(a)(ctx)
+
+    case e@in.DfltVal(_) :: ctx.Function(t) / Exclusive => dfltVal(e, t)(ctx)
+
+    case e@in.NilLit(_) :: ctx.Function(t) => dfltVal(e, t)(ctx)
+  }
+
+  private def dfltVal(e: in.Expr, fTyp: in.FunctionT)(ctx: Context): CodeWriter[vpr.Exp] =
+    ctx.expression(in.PureFunctionCall(Right(in.FunctionProxy(Names.closureDefaultFunc)(e.info)), Vector.empty, None, fTyp)(e.info))
+
+  override def statement(ctx: Context): in.Stmt ==> CodeWriter[vpr.Stmt] = default(super.statement(ctx)) {
+    case c@in.FunctionCall(_, Left(_), _, Some(_)) =>
+      specs.closureCall(c)(ctx)
+
+    case c@in.FunctionCall(targets, Right(proxy), args, Some(spec)) =>
+      val member = ctx.lookup(proxy)
+      val fType = in.FunctionT(member.args.map(_.typ), member.results.map(_.typ), Addressability.rValue)
+      val fObject = in.FunctionObject(proxy, fType)(c.info)
+      specs.closureCall(in.FunctionCall(targets, Left(fObject), args, Some(spec))(c.info))(ctx)
+
+    case c@in.MethodCall(targets, recv, meth, args, Some(spec)) =>
+      val member = ctx.lookup(meth)
+      val mType = in.FunctionT(member.args.map(_.typ), member.results.map(_.typ), Addressability.rValue)
+      val mObject = in.MethodObject(recv, meth, mType)(c.info)
+      specs.closureCall(in.FunctionCall(targets, Left(mObject), args, Some(spec))(c.info))(ctx)
+
+    case p: in.SpecImplementationProof =>
+      specImplementationProof(p)(ctx)
+  }
+
+  override def finalize(addMemberFn: vpr.Member => Unit): Unit = {
+    domain.finalize(addMemberFn)
+    specs.finalize(addMemberFn)
+    moe.finalize(addMemberFn)
+  }
+
+  /** Encodes a spec implementation proof as:
+    *
+    * if (*) {
+    *   initialize args and results
+    *   assign parameters
+    *   while(true) {
+    *     inhale precondition of spec
+    *     [ body ]
+    *     exhale postcondition of spec
+    *   }
```

Comment:
I don't understant what the problem with an infinite loop (with a `decreases _` invariant) is. What do you mean with "it does not break in the presence of inlining"? Inlining of what?

For now, I'm changing it to the following.
```
[ args, results already initialised ]
[ parameters already assigned ]
if (*) {
       while(true)
       decreases _ // assume that the loop terminates
      {
          inhale precondition of spec
         [ body ]
         exhale postcondition of spec
      }
    assume false
}
```
Changes:
- Args, results and parameters are initialised and assigned an a block enclosing the if-statement generated by the desugarer. It already worked like this, I had forgotten to reflect this in the comment.
- I added the `decreases _` invariant
- I added the `assume false` at the end
- The `*` variable is generated here instead of the desugarer

I have also added the testcase, which passes. To make it pass, I also had to add empty termination measures to implements functions, getters and conversions

Reply:
I have now changed it to the second version. The updated comment is this:

```
    [ args, results already initialised ]
    [ parameters already assigned ]
    if (*) {
      var numIterations: Int
      numIterations := proofIterator()
      while(numIterations > 0)
        decreases numIterations
      {
        inhale precondition of spec
        [ body ]
        exhale postcondition of spec
        numIterations := numIterations - 1
      }
      assume false
    }
    inhale closureImplements$[spec]([closure], [params])
```